File: webapp/go/src/isucon8/isucoin/model/setting.go
package model

import (
	"isucon8/isubank"
	"isucon8/isulogger"
	"log"

	"github.com/pkg/errors"
)

const (
	BankEndpoint = "bank_endpoint"
	BankAppid    = "bank_appid"
	LogEndpoint  = "log_endpoint"
	LogAppid     = "log_appid"
)

//go:generate scanner
type Setting struct {
	Name string
	Val  string
}

func SetSetting(d QueryExecutor, k, v string) error {
	_, err := d.Exec(`INSERT INTO setting (name, val) VALUES (?, ?) ON DUPLICATE KEY UPDATE val = VALUES(val)`, k, v)
	return err
}

func GetSetting(d QueryExecutor, k string) (string, error) {
	s, err := scanSetting(d.Query(`SELECT * FROM setting WHERE name = ?`, k))
	if err != nil {
		return "", err
	}
	return s.Val, nil
}

func Isubank(d QueryExecutor) (*isubank.Isubank, error) {
	ep, err := GetSetting(d, BankEndpoint)
	if err != nil {
		return nil, errors.Wrapf(err, "getSetting failed. %s", BankEndpoint)
	}
	id, err := GetSetting(d, BankAppid)
	if err != nil {
		return nil, errors.Wrapf(err, "getSetting failed. %s", BankAppid)
	}
	return isubank.NewIsubank(ep, id)
}

func Logger(d QueryExecutor) (*isulogger.Isulogger, error) {
	ep, err := GetSetting(d, LogEndpoint)
	if err != nil {
		return nil, errors.Wrapf(err, "getSetting failed. %s", LogEndpoint)
	}
	id, err := GetSetting(d, LogAppid)
	if err != nil {
		return nil, errors.Wrapf(err, "getSetting failed. %s", LogAppid)
	}
	return isulogger.NewIsulogger(ep, id)
}

func sendLog(d QueryExecutor, tag string, v interface{}) {
	logger, err := Logger(d)
	if err != nil {
		log.Printf("[WARN] new logger failed. tag: %s, v: %v, err:%s", tag, v, err)
		return
	}
	err = logger.Send(tag, v)
	if err != nil {
		log.Printf("[WARN] logger send failed. tag: %s, v: %v, err:%s", tag, v, err)
	}
}

File: webapp/php/src/Isucon8/Isucoin/Model/Setting.php
<?php

use Isucon8\PDOWrapper;
use Isucon8\PDOWrapperTxn;
use Isucon8\Isubank\Isubank;
use Isucon8\Isulogger\Isulogger;

const BANK_ENDPOINT = 'bank_endpoint';
const BANK_APPID    = 'bank_appid';
const LOG_ENDPOINT  = 'log_endpoint';
const LOG_APPID     = 'log_appid';

function SetSetting(PDOWrapperTxn $txn, string $k, string $v): void {
    $txn->execute('INSERT INTO setting (name, val) VALUES (?, ?) ON DUPLICATE KEY UPDATE val = VALUES(val)', $k, $v);
}

function GetSetting(PDOWrapper $dbh, string $k): string {
    return $dbh->select_one('SELECT val FROM setting WHERE name = ?', $k);
}

function Isubank(PDOWrapper $dbh): Isubank {
    $ep = GetSetting($dbh, BANK_ENDPOINT);
    $id = GetSetting($dbh, BANK_APPID);
    return new Isubank($ep, $id);
}

function Logger(PDOWrapper $dbh): Isulogger {
    $ep = GetSetting($dbh, LOG_ENDPOINT);
    $id = GetSetting($dbh, LOG_APPID);
    return new Isulogger($ep, $id);
}

function sendLog(PDOWrapper $dbh, string $tag, array $v): void {
    $logger = null;
    try {
        $logger = Logger($dbh);
    } catch(\Throwable $throwable) {
        error_log(sprintf('[WARN] new logger failed. tag: %s, v: %s, err:%s', $tag, serialize($v), $throwable));
    }
    if ($logger === null) {
        return;
    }
    try {
        $logger->send($tag, $v);
    } catch(\Throwable $throwable) {
        error_log(sprintf('[WARN] logger send failed. tag: %s, v: %s, err:%s', $tag, serialize($v), $throwable));
    }
}
